### PR TITLE
Require captcha verification when joining section through /join/:section_code if more than 3 section join attempts in 24 hours

### PR DESCRIPTION
**Pull request by drew-beckmen**
This pull request represents the reintroduction of the code from #38026 that was later reverted. Issues with pull request #38026 are well documented in this [Slack thread](https://codedotorg.slack.com/archives/C0T0PNTM3/p1606971655144400) as well as this [COE report](document/d/14Ul9OGCgw5XmpnYZv0Q7X5x3baAU2oPMeflQce1jUGQ/edit#). The Honeybadger errors generated by #38026 are [here](https://app.honeybadger.io/projects/3240/faults/67731310).

## Improvements over #38026 

- The issue last time was the use of `save(validate: false)` when new users were being persisted to the database for the first time. I've removed all instances of `save(validate: false)`. 
- There are two instances in which I use `save` to persist to the database: `increment_section_attempts` and `reset_section_attempts_if_needed`. For `increment_section_attempts`, the method is only invoked on `current_user` or `@user` where the command`@user.save` has already been executed. `reset_section_attempts_if_needed` is called in `display_captcha?`, and the `display_captcha?` method is also only even called on `current_user`. Thus, I don't expect my new logic to interfere with the pre-existing process of getting new entries into the `users` table.
- Added some basic disabling/enabling of submit buttons in the forms rendered by `student_user_new.html.haml` to try and prevent users from submitting bad data in the first place. When the captcha verification is required, I also won't let users press submit without filling out the captcha. The only exception is if the captcha verification times out, the submit button will not disable itself again. If the user submits the form after the verification timed out, they'll receive a `flash` message that CAPTCHA is required.

## Future Work

Add same process to join section workflow through the teacher/student dashboard. Some of that work has already been completed in PR #38032 

## Links

- [Documentation](https://docs.google.com/document/d/1OeJwk8mx5BrovwdC74bBvkSnphxek3UpRkI3zNytSQ4/edit#)
- [jira](https://codedotorg.atlassian.net/browse/LP-1644?atlOrigin=eyJpIjoiNjMwNjQyODI5Y2Q4NDlkYmE5MjNlODlkNzg5OWQ0MTUiLCJwIjoiaiJ9)
- [jira - Epic](https://codedotorg.atlassian.net/browse/LP-1378?atlOrigin=eyJpIjoiNjYyYmJiY2YyMzlhNGM5MDkzNWE4ODFlMzcxOWE2NjEiLCJwIjoiaiJ9)

## Testing story

Added unit tests to verify new logic in `user.rb` and `followers_controller.rb`. Prior to reintroducing these changes, I also expanded the test coverage surrounding the join section workflow in PR #38225 - the changes in #38225 would have caught the issue in #38026 

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/followers_controller.rb
Comment (cforkish) [inline dashboard/app/controllers/followers_controller.rb]: if `@section` is always defined, can't you just use `@section.code` here and omit the previous line?

File: dashboard/app/models/user.rb
Comment (cforkish) [inline dashboard/app/models/user.rb]: does `DateTime`arithmetic operate on "day units"? to me, `(DateTime.now - DateTime.parse(properties['section_attempts_last_reset'])).to_i > 0` looks like you're checking to see if any time has passed at all, so if that difference is operating on day units, could you add that to the comment? if not, i think you want something other than 0 here (depending on the units)

File: dashboard/app/controllers/followers_controller.rb
Comment (cforkish) [inline dashboard/app/controllers/followers_controller.rb]: NIT: imo `current_user?.increment_section_attempts` is cleaner

File: dashboard/app/models/user.rb
Comment (cforkish) [inline dashboard/app/models/user.rb]: this comment is a little confusing to me, because it sounds like you're about to do that (make sure user is registered), but the code that follows doesn't appear to do it.

are we accomplishing that by checking for `current_user` before calling this? (i'm not sure if having a `current_user` means they are registered). if that's the case, it seems like the fact that this is an instance method means it could only be called on a registered user, in which case i'd just remove this comment.

if we're verifying that the user is registered somewhere else, i'd suggest either moving that verification into this meothd, or at least explaining how/where we're doing it in this comment.

**Review by cforkish — CHANGES_REQUESTED** [review]
overall this looks great, but i have left a few questions i'd like answers to before i approve.

**Comment by cforkish** [comment]
one other general comment that is not a blocker for me: our style guide specifies that we should do our best to stick to an 80-char line limit. this is not commonly adhered to in our ruby code, but imo it should be. up to you, but i'd greatly appreciate it!

File: dashboard/app/models/user.rb
Comment (drew-beckmen) [reply dashboard/app/models/user.rb]: I think this was a comment that I left to myself but did not remove. Should have labeled it with `TODO` — the issue last time was calling methods in the user model on `@user` instances that had not yet been persisted to the database, so I wanted to be extra careful that doesn’t happen again by tracing all the method calls in which `save` Is invoked.

Comment (drew-beckmen) [reply dashboard/app/models/user.rb]: Also would love to get confirmation that `current_user` does in fact always reference a User object that has already been persisted to the database. The tricky thing about Rails is that instance methods can be called on objects before they ever go into the database. Only a `save` call fire of the SQL necessary to persist any changes.

File: dashboard/app/models/user.rb
Comment (jmkulwik) [inline dashboard/app/models/user.rb]: NIT typo: "aare"

File: dashboard/app/views/followers/student_user_new.html.haml
Comment (jmkulwik) [inline dashboard/app/views/followers/student_user_new.html.haml]: What does `id: 'register_join_section'` get used for?

File: dashboard/app/views/followers/student_user_new.html.haml
Comment (jmkulwik) [inline dashboard/app/views/followers/student_user_new.html.haml]: `id: 'join_section'` seems like it may be a common id name. Can you do a quick sanity check of this page to make sure this id isn't used anywhere else? (or if you can make it a class instead, that would be preferred - not necessary though.)

File: dashboard/test/models/user_test.rb
Comment (jmkulwik) [inline dashboard/test/models/user_test.rb:4462]: Your tests look great!

File: dashboard/app/controllers/followers_controller.rb
Comment (daynew) [inline dashboard/app/controllers/followers_controller.rb:36]: Why retry on `RecordNotUnique`? That doesn't sound like an error which will be fixed by retrying.

File: dashboard/app/models/user.rb
Comment (daynew) [inline dashboard/app/models/user.rb]: What is `properties` supposed to be used for? Why is the code using `properties` instead of adding a new attribute to the `User` model?

Comment (bethanyaconnor) [reply dashboard/app/models/user.rb]: Most new fields of models are put into `properties`. The fields in `properties` of the User model are [here](https://github.com/code-dot-org/code-dot-org/blob/0cfe36263ee85b4581622fe928ce3c2f413b347e/dashboard/app/models/user.rb#L96). I think `section_attempts` makes sense in `properties`, unless there's a reason to query for it directly.

Comment (bethanyaconnor) [reply dashboard/app/models/user.rb]: But on that note, `section_attempts` and `section_attempts_last_reset` should be added to that last I linked

File: dashboard/app/models/user.rb
Comment (bethanyaconnor) [reply dashboard/app/models/user.rb]: I'm not sure if `current_user` is always persisted but you could always check [`current_user.persisted?`](https://apidock.com/rails/ActiveRecord/Persistence/persisted%3F)

File: dashboard/app/models/user.rb
Comment (bethanyaconnor) [reply dashboard/app/models/user.rb]: Which also should allow you to reference them directly instead of through `properties`, which might've been what Dayne was going for all along....

(Sorry for multiple comments, it's still early so my brain is still booting)

File: dashboard/app/models/user.rb
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: Is this comment still accurate?  Does nil or 0 get returned when there are no section join attempts?

File: dashboard/app/controllers/followers_controller.rb
Comment (jamescodeorg) [inline dashboard/app/controllers/followers_controller.rb]: Do we end up here just for the POST request or both the GET and POST requests?  (Want to double-check that we're not double counting attempts.)

File: dashboard/app/controllers/followers_controller.rb
Comment (jamescodeorg) [reply dashboard/app/controllers/followers_controller.rb:36]: I think this code was already there before (but was indented).  You can see just the code that Drew added by hiding whitespace changes.

File: dashboard/app/models/user.rb
Comment (jamescodeorg) [inline dashboard/app/models/user.rb:2280]: I'm not really sure how to fix this, but display_captcha? has a potentially unexpected side effect of saving the user object (which in turn triggers other code).  At the very least, this should be mentioned in the header comment, but perhaps there's a way to structure the code to make this function more side-effect-free?

File: dashboard/app/models/user.rb
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: Why do we need to reset section attempts if it nil/0?  (Also, depending on your answer to my other question above, we may not need to check for nil here.)

File: dashboard/app/models/user.rb
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: I'm wondering if here (and on 2279) we should save the object only if had previously been persisted.

One of the contributing factors in the original implementation is that `display_captcha?` eventually called `save(validate: false)` and that ended up being the first time the object was validated and persisted.  Because of the way the user model is written, some hooks only get run on the initial save.  It would be lower risk if this change did not change when that initial save occurs.

**Review by jamescodeorg — COMMENTED** [review]
Some of the comments I left are a bit more paranoid than I would normally be, but this area has been proven to be tricky.  :-)  Let me know if you want to chat or jump on a call to discuss anything.

File: dashboard/app/models/user.rb
Comment (daynew) [reply dashboard/app/models/user.rb]: @bethanyaconnor I didn't know what you were talking about so thank you for the explanation. I didn't see any other direct uses of the 'properties' field and that field has no documentation, so it seemed weird to me. Is this a common active record pattern? Does it have a name I could look up and learn more about it?

Comment (bethanyaconnor) [reply dashboard/app/models/user.rb]: @daynew Yeah sorry for the extra comments 😬  I believe this is a custom thing defined [here](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/concerns/serialized_properties.rb).

This is definitely one of those things that I didn't know existed until I had to add a new attribute last summer

File: dashboard/app/controllers/followers_controller.rb
Comment (drew-beckmen) [reply dashboard/app/controllers/followers_controller.rb]: Yep definitely -- I think I included this more a safeguard in case my Rails knowledge failed me, but I believe the declaration of `@section` in `load_section` creates an instance variable that is accessible globally.

File: dashboard/app/controllers/followers_controller.rb
Comment (drew-beckmen) [reply dashboard/app/controllers/followers_controller.rb:36]: Yes according to the git blame, the retryable block was included in the original version of this code.

File: dashboard/app/models/user.rb
Comment (drew-beckmen) [reply dashboard/app/models/user.rb]: wow, didn't know about this either. thanks all for enlightening me --- this will definitely clean up the code

File: dashboard/app/models/user.rb
Comment (drew-beckmen) [reply dashboard/app/models/user.rb]: Thanks for pointing it out. I changed the method to return zero in the case there are no section attempts. Felt cleaner and safer than having `nil` values floating around. I've updated the comment.

File: dashboard/app/models/user.rb
Comment (drew-beckmen) [reply dashboard/app/models/user.rb]: Subtracting DateTime objects in Ruby gives the difference in days as a floating point number: https://stackoverflow.com/questions/30953974/datetime-subtraction-in-ruby-2

The key here is the `to_i` casting. If the two objects are less than 1 day apart, the subtraction will yield a fractional number < 1. By casting to an int, we can check to ensure at least 1 full day has passed. Does this make sense? I'll work to clarify with an inline comment.

Comment (drew-beckmen) [reply dashboard/app/models/user.rb]: Agreed completely. Given `num_section_attempts` will always return 0 if the value is `nil`, we don't need to check for `nil` values here.

File: dashboard/app/models/user.rb
Comment (drew-beckmen) [reply dashboard/app/models/user.rb:2280]: I've really struggled with this question myself. It's a very good question, and I would love to maybe brainstorm together on it. The issue I have is a user could join a section for the first time and when we check whether or not to display the captcha, all of the properties related to joining a section would return `nil`. We could probably account for that case, but then there's the question of a user joining a section for the 4th time in 25 hours. In the 25th hour, we'll need to reset the values before determining whether or not to display the captcha.

Maybe `increment_section_attempts` is the only place we call `save` and display_captcha can reset section attempts where necessary and return the resulting boolean without actually changing values in the db. Then, when we actually account for the fact that they pressed "Submit" on joining a section, we can handle any reseting. That was a lot... let me know if it made sense or if you would like to discuss it further.

File: dashboard/app/views/followers/student_user_new.html.haml
Comment (drew-beckmen) [reply dashboard/app/views/followers/student_user_new.html.haml]: Leftover from the changes I refactored out of this PR to include some client-side validations on the form. Will remove

File: dashboard/app/controllers/followers_controller.rb
Comment (drew-beckmen) [reply dashboard/app/controllers/followers_controller.rb]: Is this valid syntax? I would expect an error to the tune of method `current_user?` doesn't exist

File: dashboard/app/models/user.rb
Comment (drew-beckmen) [reply dashboard/app/models/user.rb]: Started a discussion about the `current_user` method inherited from `ApplicationController` in this [Slack thread](https://codedotorg.slack.com/archives/C0T10HG6N/p1611198096002600).

File: dashboard/app/models/user.rb
Comment (drew-beckmen) [reply dashboard/app/models/user.rb:2280]: Make it a pure function - want it to be clear that it has no side effects.
Try not to alter state in a method returning a boolean:
1) if reset was over a day ago, return false; then, 
2) if number of attempts is greater than or equal to three, return true; else, return false.

Then, expand code in the increment method and note that it should only be called on a user object that has been persisted –  can add this check explicitly.

File: dashboard/app/controllers/followers_controller.rb
Comment (drew-beckmen) [inline dashboard/app/controllers/followers_controller.rb]: from a functional point of view, this is not useful. The user is already over the limit.

File: dashboard/app/models/user.rb
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: I think we can probably remove this comment now that the function is defensive about saving only if the user object was previously persisted?

File: dashboard/app/models/user.rb
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: It'd probably be useful for this comment to indicate _why_.

File: dashboard/app/models/user.rb
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: nit: I think this might be a bit clearer if this said `since last reset` instead of `since last attempt`

File: dashboard/app/models/user.rb
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: nit: I think this might be a bit clearer if this said `since last reset` instead of `since last section attempt`

**Review by jamescodeorg — APPROVED** [review]
Some minor feedback about comments but generally looks good to me.  Since we worked on the last set of changes together, I would wait for at least one more approval before merging.

**Review by jmkulwik — APPROVED** [review]
Great work on this! Looks good to me, but deferring to James or Charlie who have been following more closely.

File: dashboard/app/models/user.rb
Comment (bethanyaconnor) [inline dashboard/app/models/user.rb]: It looks like any error here will be silently swallowed.  `save` doesn't raise an error if the save was unsuccessful and instead returns a boolean, whereas `save!` does throw an exception. I'd recommend either using `save!` or using the return value of this function

**Pull request merged**
